CenterForOpenScience/waterbutler
# [SVCS-552] Remove unneeded API calls in folder_file_ops

Added construct_path and construct_empty_path to base-
provider for individual providers to make paths for folder-file-ops
without needing to make API calls







## Ticket
https://openscience.atlassian.net/browse/SVCS-552


## Purpose
the Baseprovider class `folder_file_op` function was making calls to get metadata, and then making the same calls again when calling revalidate path. It is possible to create paths from this metadata and forgo making more API calls.



## Changes
Added 2 new functions to the Baseprovider. `construct_path` and `construct_empty_path`. 
Construct_path takes the parent path and metadata. Each provider then has its own implementation on how to get a path from this. Some will simply use it as a proxy to `revalidate_path` since their `revalidate_path` does not make API calls. If an individual providers `revalidate_path` does make api calls, it will instead call `path_from_metadata` or something similar.

`construct_empty_path` returns a math from metadata where the identifier is none( ie it doesnt actually exit at that location). 

Added tests to test these new functions. `construct_empty_path` is tested in core/test_provider since almost all providers use the base definition.

`construct_path` is tested individually in providers and is ALWAYS tested against the output of `revalidate_path` because that is what it is replacing.

Dataverse is the only provider that could still makes API calls here, however I am fairly certain that dataverse cannot perform `folder_file_ops` so it should never call those functions.



## Side effects

There could be some timing issues, however this seems unlikely



## QA Notes

To test, move or copy a folder containing at least one file across 2 providers.
ie move folder1 from OSFstorage to GoogleDrive. It should work and make less API calls

Figshare may not work quite correctly. It is very hard to test with the many bugs figshare has.



## Deployment Notes
